src/main/java/gregtech/common/covers/filter/FilterTypeRegistry.java

```diff
@@ -74,30 +88,53 @@ public static FluidFilter createFluidFilterById(int filterId) {
         return createNewFilterInstance(filterClass);
     }
 
-    public static ItemFilter getItemFilterForStack(ItemStack itemStack) {
-        Integer filterId = itemFilterIdByStack.get(new ItemAndMetadata(itemStack));
-        if (filterId == null) {
-            return null;
+    public static @NotNull ItemFilter getItemFilterForStack(ItemStack itemStack) {
+        int filterId = getFilterIdForStack(itemStack);
+        if (filterId == -1) {
+            throw new IllegalArgumentException(
+                    String.format("Failed to create filter instance for stack %s", itemStack));
         }
         Class<? extends ItemFilter> filterClass = itemFilterById.get(filterId);
-        return createNewFilterInstance(filterClass);
+        return createNewFilterInstance(filterClass, itemStack);
     }
 
-    public static FluidFilter getFluidFilterForStack(ItemStack itemStack) {
-        Integer filterId = fluidFilterIdByStack.get(new ItemAndMetadata(itemStack));
-        if (filterId == null) {
-            return null;
+    public static @NotNull FluidFilter getFluidFilterForStack(ItemStack itemStack) {
+        int filterId = getFilterIdForStack(itemStack);
+        if (filterId == -1) {
+            throw new IllegalArgumentException(
+                    String.format("Failed to create filter instance for stack %s", itemStack));
         }
         Class<? extends FluidFilter> filterClass = fluidFilterById.get(filterId);
-        return createNewFilterInstance(filterClass);
+        return createNewFilterInstance(filterClass, itemStack);
     }
 
-    private static <T> T createNewFilterInstance(Class<T> filterClass) {
+    public static int getFilterIdForStack(ItemStack stack) {
+        int id = -1;
+        if (isItemFilter(stack)) id = itemFilterIdByStack.getOrDefault(new ItemAndMetadata(stack), -1);
+        else if (isFluidFilter(stack)) id = fluidFilterIdByStack.getOrDefault(new ItemAndMetadata(stack), -1);
+        return id;
+    }
+
+    private static <T> @NotNull T createNewFilterInstance(Class<T> filterClass, ItemStack stack) {
         try {
-            return filterClass.newInstance();
+            return filterClass.getDeclaredConstructor(stack.getClass()).newInstance(stack);
```

Comment:
I don't like this approach of using reflection for filters. If this class is actually necessary, I think moving to some functional interface to create new filter instances would be preferable.

Reply:
could probably be replaced with a method reference of the constructor